### ISSUE TITLE
Lack of variation causes segmentation fault with polyphase

### ISSUE THREAD

**Issue by peterdfields** [whatshap/whatshap#441]
I'm presently using whatshap on a use case which might not be so common, specifically Illumina sequencing of a sequence capture library. Here the 'reference' is a de novo assembly of each of these captured sequences/genes, many of which are quite small. What seems to be happening is that in at least some cases where entries in the VCF exist which are non-reference, but invariant (so something like 1/1/1/1 in the GT field for a tetraploid), result in a segmentation fault as in the following:

```
======== Working on chromosome 'gene_100_consensus'
---- Processing individual unknown
Number of variants skipped due to missing genotypes: 0
Number of remaining heterozygous variants: 0
Found 0 reads covering 0 variants
Kept 0 reads that cover at least two variants each
Detecting connected components with weak interconnect ..
Segmentation fault (core dumped)
```

When I remove these entries from the VCF with bcftools whatshap polyphase completes successfully. I have checked this invariant non-reference issue on diploid samples (so `whatshap phase`) and the run completes successfully, so the issue seems specific to `polyphase`.  The error arises whether one specifies `--indels` or not, and across multiple versions of the software. For now I'll just remove this sites with bcftools. I can provide data if it would be helpful to reproduce the described error. Please let me know if additional information would be helpful.

**Comment by schrins** [comment]
It is already peculiar, that no reads and no variants are detected. Does this also happen, when you run it without the non-ref homozygous entries? In principle, such entries should not make a difference at all, because they will be discarded before reading the alignments.

I can reproduce a segmentation fault, if the core algorithm is started with zero reads or zero variants.

**Comment by peterdfields** [comment]
If I remove the sites which are homozygous non-ref (1/1/1/1 in the case of a tetraploid) with bcftools then `whatshap polyphase` can run without causing a segmentation fault.

**Comment by schrins** [comment]
But does it report 0 reads covering 0 variants as well, then?

It is an issue that the program crashes with 0 reads and variants, but if it wouldn't crash, the result would be an empty phasing (i.e. all variants left unphased) anyways. That's why I'm bothered with the number of detected reads/variants. It looks like there is another issue besides from the segfault.

**Comment by peterdfields** [comment]
Do you mean in the bam file or the output to the terminal by whatshap? I also see another type of odd error, where again there is only a single variant across the full assembled probe (so no phasing is necessary/possible), but a seg fault happens:

```
======== Working on chromosome '3GT_maker_scaffold_1335_augustus_gene_0.3_mRNA_1_tr'
---- Processing individual sample
Number of variants skipped due to missing genotypes: 0
Number of remaining heterozygous variants: 1
Found 445 reads covering 1 variants
Kept 0 reads that cover at least two variants each
Detecting connected components with weak interconnect ..
Segmentation fault (core dumped)
```

If I remove these entries from the VCF the run continues successfully but it's a bit inconvenient to cut up the vcf this way.

**Comment by schrins** [comment]
With the previous post I mean: If you take the run from your first post, but you remove the 1/1/1/1-entries from the VCF. Does Whatshap still only find 0 reads and variants in the terminal output?

If you want, you can also share some example data that produces an error. I would have a look at it and see what is going on inside the code.

**Comment by schrins** [comment]
The branch now exists: `issue-441`

If you didn't already work with branches, I recently wrote some instructions how to clone and combine a feature branch directly from the repository: https://github.com/whatshap/whatshap/issues/439#issuecomment-1472170854

**Comment by peterdfields** [comment]
I was able to pull the `issue-441` branch and build the tool successfully. Many of the samples for which the error described above occurred can now run to completion. However, a subset of samples are failing with a different error:

```
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "/home/peter/miniconda3/envs/whtest/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/peter/bioinformatics/whatshap/whatshap/polyphase/algorithm.py", line 264, in phase_single_block_mt
    result = phase_single_block(block_id, submatrix, genotype_slice, subphasing, param, timers)
  File "/home/peter/bioinformatics/whatshap/whatshap/polyphase/algorithm.py", line 213, in phase_single_block
    res = solve_polyphase_instance(subm, subgeno, sub_param, timers, quiet=True)
  File "/home/peter/bioinformatics/whatshap/whatshap/polyphase/algorithm.py", line 80, in solve_polyphase_instance
    phase_single_block(
  File "/home/peter/bioinformatics/whatshap/whatshap/polyphase/algorithm.py", line 186, in phase_single_block
    threads, haplotypes = run_threading(
  File "/home/peter/bioinformatics/whatshap/whatshap/polyphase/threading.py", line 42, in run_threading
    affine_switch_cost = ceil(compute_readlength_snp_distance_ratio(allele_matrix) / 1.0)
  File "/home/peter/bioinformatics/whatshap/whatshap/polyphase/threading.py", line 69, in compute_readlength_snp_distance_ratio
    return length / len(allele_matrix)
ZeroDivisionError: division by zero
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/peter/miniconda3/envs/whtest/bin/whatshap", line 8, in <module>
    sys.exit(main())
  File "/home/peter/bioinformatics/whatshap/whatshap/__main__.py", line 64, in main
    module.main(args)
  File "/home/peter/bioinformatics/whatshap/whatshap/cli/polyphase.py", line 547, in main
    run_polyphase(**vars(args))
  File "/home/peter/bioinformatics/whatshap/whatshap/cli/polyphase.py", line 250, in run_polyphase
    ) = phase_single_individual(
  File "/home/peter/bioinformatics/whatshap/whatshap/cli/polyphase.py", line 318, in phase_single_individual
    result = solve_polyphase_instance(allele_matrix, genotype_list, param, timers, prephasing)
  File "/home/peter/bioinformatics/whatshap/whatshap/polyphase/algorithm.py", line 112, in solve_polyphase_instance
    blockwise_results = [res.get() for res in process_results]
  File "/home/peter/bioinformatics/whatshap/whatshap/polyphase/algorithm.py", line 112, in <listcomp>
    blockwise_results = [res.get() for res in process_results]
  File "/home/peter/miniconda3/envs/whtest/lib/python3.10/multiprocessing/pool.py", line 774, in get
    raise self._value
ZeroDivisionError: division by zero
```
I attempted some small modifications to command line flags, for example removing `--indels`, not specifying `--threads`, but the error persists. Please let me know if any additional information would be helpful in order to troubleshoot this issue.

**Comment by peterdfields** [comment]
Should I move this new error to another issue and close this issue given the lack of diversity may not be causal for the observed error?

**Comment by schrins** [comment]
No, I think we can resolve this here as well. I did not find the time to look at this before.

In the offending line a division by the size (i.e. read count) of the current allele matrix is done. However, after looking where this matrix object comes from, I don't really see how it can be empty here. Can you share the data that causes this issue? This would be the easiest way, I think.

**Comment by peterdfields** [comment]
Okay, I can share the data. Is there an email I can share the google drive link with?

**Comment by schrins** [comment]
Yes, if you click on my profile, my email address is on the left.

**Comment by schrins** [comment]
Thanks for sharing the data. I was able to reproduce and fix the crash. However, I noticed a few other oddities, I didn't see before. I don't know how malicious they are for your results, maybe they are just artifacts. I will create another issue for this, because I don't know right away how to fix them.

**Comment by peterdfields** [comment]
Thank you for having a look. Once you have an opportunity to clarify the type of oddities you're noticing I can try to design an analysis to determine if they might have downstream issues for the analysis I'm working on. I'm not sure if the error you're looking at now is related but for octoploids I see the following error message:

```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/peter/miniconda3/envs/whtest/bin/whatshap", line 8, in <module>
    sys.exit(main())
  File "/home/peter/bioinformatics/whatshap/whatshap/__main__.py", line 64, in main
    module.main(args)
  File "/home/peter/bioinformatics/whatshap/whatshap/cli/polyphase.py", line 547, in main
    run_polyphase(**vars(args))
  File "/home/peter/bioinformatics/whatshap/whatshap/cli/polyphase.py", line 250, in run_polyphase
    ) = phase_single_individual(
  File "/home/peter/bioinformatics/whatshap/whatshap/cli/polyphase.py", line 318, in phase_single_individual
    result = solve_polyphase_instance(allele_matrix, genotype_list, param, timers, prephasing)
  File "/home/peter/bioinformatics/whatshap/whatshap/polyphase/algorithm.py", line 112, in solve_polyphase_instance
    blockwise_results = [res.get() for res in process_results]
  File "/home/peter/bioinformatics/whatshap/whatshap/polyphase/algorithm.py", line 112, in <listcomp>
    blockwise_results = [res.get() for res in process_results]
  File "/home/peter/miniconda3/envs/whtest/lib/python3.10/multiprocessing/pool.py", line 774, in get
    raise self._value
pulp.apis.core.PulpSolverError: Pulp: Error while trying to execute, use msg=True for more details/home/peter/miniconda3/envs/whtest/lib/python3.10/site-packages/pulp/solverdir/cbc/linux/64/cbc
```

**Comment by schrins** [comment]
I resolved a PuLP problem in #439 because the package seems to have different solvers available on different systems, even if version and installation method were the same. I will work on that issue next, so I can merge those changes into the issue-441 branch once it is done. It is strange though, that this error did not occur earlier. Did it take suspiciuosly long to process or a huge amount of RAM? There are some parts in the algorithm that scale exponentially with ploidy (including the linear program that is solved here by PuLP) and since I never ran WhatsHap with more than ploidy 6 on some test data, I don't know how the algorithm behaves with ploidy 8.

**Comment by peterdfields** [comment]
In the case of the ploidy 8 it did take quite a long time before the error arose compared to even ploidy 6 and especially 4 and 2. The machine the job is running on has 128Gb of memory so I don't think it ran out of memory but I didn't monitor the memory use. I can re-run the job and monitor resource use more closely if that would be helpful?

**Comment by schrins** [comment]
You can do this, but it would be more helpful, if the solver is selected correctly. You could either wait until #439 is resolved and merged, or you can change line 433 in `whatshap/polyphase/reorder.py`

from 

```
solver = PULP_CBC_CMD(msg=0)
```

to

```
solver = solver = get_ilp_solver()
```

and add the following code at the top of the file (after the current imports):

```
from pulp import listSolvers, getSolver

def get_ilp_solver():
    solvers = listSolvers(onlyAvailable=True)
    prio = ["GUROBI_CMD", "GUROBI", "COIN_CMD", "PULP_CBC_CMD"]
    for name in prio:
        if name in solvers:
            return getSolver(name, msg=0)
    return getSolver(solvers[0], msg=0)
```
I don't want to push these changes myself to this branch, because this would just create conflicts with the other branch.

**Comment by peterdfields** [comment]
I made the modifications and restarted the run on one of the octoploid samples. The run is still going after nearly 24 hours but still has many molecules to go. I will update if an error arises as before.

**Comment by peterdfields** [comment]
An error on the ploidy 8 sample popped up again even with the suggested modifications:
```
Traceback (most recent call last):
  File "/home/peter/miniconda3/envs/wh_octo/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/peter/bioinformatics/whatshap_rebuild/whatshap/polyphase/algorithm.py", line 264, in phase_single_block_mt
    result = phase_single_block(block_id, submatrix, genotype_slice, subphasing, param, timers)
  File "/home/peter/bioinformatics/whatshap_rebuild/whatshap/polyphase/algorithm.py", line 225, in phase_single_block
    run_reordering(allele_matrix, clustering, threads, haplotypes, breakpoints, prephasing)
  File "/home/peter/bioinformatics/whatshap_rebuild/whatshap/polyphase/reorder.py", line 184, in run_reordering
    perms = get_optimal_permutations(breakpoints, lllh, ploidy, aff)
  File "/home/peter/bioinformatics/whatshap_rebuild/whatshap/polyphase/reorder.py", line 443, in get_optimal_permutations
    model.solve(solver)
  File "/home/peter/miniconda3/envs/wh_octo/lib/python3.10/site-packages/pulp/pulp.py", line 1913, in solve
    status = solver.actualSolve(self, **kwargs)
  File "/home/peter/miniconda3/envs/wh_octo/lib/python3.10/site-packages/pulp/apis/coin_api.py", line 137, in actualSolve
    return self.solve_CBC(lp, **kwargs)
  File "/home/peter/miniconda3/envs/wh_octo/lib/python3.10/site-packages/pulp/apis/coin_api.py", line 206, in solve_CBC
    raise PulpSolverError(
pulp.apis.core.PulpSolverError: Pulp: Error while trying to execute, use msg=True for more details/home/peter/miniconda3/envs/wh_octo/lib/python3.10/site-packages/pulp/solverdir/cbc/linux/64/cb

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/peter/miniconda3/envs/wh_octo/bin/whatshap", line 8, in <module>
    sys.exit(main())
  File "/home/peter/bioinformatics/whatshap_rebuild/whatshap/__main__.py", line 64, in main
    module.main(args)
  File "/home/peter/bioinformatics/whatshap_rebuild/whatshap/cli/polyphase.py", line 547, in main
    run_polyphase(**vars(args))
  File "/home/peter/bioinformatics/whatshap_rebuild/whatshap/cli/polyphase.py", line 250, in run_polyphase
    ) = phase_single_individual(
  File "/home/peter/bioinformatics/whatshap_rebuild/whatshap/cli/polyphase.py", line 318, in phase_single_individual
    result = solve_polyphase_instance(allele_matrix, genotype_list, param, timers, prephasing)
  File "/home/peter/bioinformatics/whatshap_rebuild/whatshap/polyphase/algorithm.py", line 112, in solve_polyphase_instance
    blockwise_results = [res.get() for res in process_results]
  File "/home/peter/bioinformatics/whatshap_rebuild/whatshap/polyphase/algorithm.py", line 112, in <listcomp>
    blockwise_results = [res.get() for res in process_results]
  File "/home/peter/miniconda3/envs/wh_octo/lib/python3.10/multiprocessing/pool.py", line 774, in get
    raise self._value
pulp.apis.core.PulpSolverError: Pulp: Error while trying to execute, use msg=True for more details/home/peter/miniconda3/envs/wh_octo/lib/python3.10/site-packages/pulp/solverdir/cbc/linux/64/cbc
```

Please let me know if any additional information would be helpful to diagnose the issue.

**Comment by schrins** [comment]
The only thing I could think of, would be to follow the error message's advice and set the `msg`-parameter to 1 in the code snippet I posted above. This will probably create a lot of terminal output from the PuLP package (which is why we disabled it by default).

**Comment by peterdfields** [comment]
Okay, I will give this a try. Just so I'm sure I've made the modifications correctly, here is the top 20+ lines of `whatshap/polyphase/reorder.py`

```
import itertools as it
import logging
from collections import defaultdict
from bisect import bisect_right
from typing import List
from math import log, exp
from functools import reduce
from operator import mul
from copy import deepcopy
from pulp import LpProblem, LpVariable, LpMaximize, LpInteger, PULP_CBC_CMD

from whatshap.polyphase import PolyphaseBlockResult, PhaseBreakpoint
from pulp import listSolvers, getSolver

def get_ilp_solver():
    solvers = listSolvers(onlyAvailable=True)
    prio = ["GUROBI_CMD", "GUROBI", "COIN_CMD", "PULP_CBC_CMD"]
    for name in prio:
        if name in solvers:
            return getSolver(name, msg=1)
        return getSolver(solvers[0], msg=1)

logger = logging.getLogger(__name__)
```

I'll restart the run now.

**Comment by schrins** [comment]
The code changes look correct. I might have "found" the issue in a different context, see issue #449 .

My feeling is, that this will be not be easy to fix, because it looks like an algorithmic issue rather than anything code-related. Any instances crashing on `reorder.py` with ploidy > 4 are likely to be affected by the same problem.

**Comment by peterdfields** [comment]
In case it's useful I created a gist of the full log of the run prior to the changes you made today [here](https://gist.github.com/peterdfields/ab2eef49fc0fdb4e4f6d7a7eb8ec7780). I will pull the changes you made to the branch `issue-441` and re-run the samples that caused problems for ploidy 4N, 6N, and 8N. I should have updates for the first two on Friday and at the beginning of the week for last one. Thank you again for taking the time to try to solve this issue.

**Comment by schrins** [comment]
Sorry, I overlooked the log you provided. It can be seen that the last ILP in the log has over 1.8 million rows and more than 5 million entries. In some testing I made for issue #449 an ILP with about 5% of this size already took an hour to solve. For me it is pretty obvious that the large ILP is the problem here and that your computer (and likely any computer in the world) could never finish the calculation and run out of memory at some points.

Did any other issues show up? Otherwise, I would refer to #449 for the octoploid issue and close this one.

**Comment by peterdfields** [comment]
I can confirm that the issues which arose for ploidy 4 and 6 were resolved with the #439 modifications. I will close this issue for now and give the modifications in #449 a try for the ploidy > 6 samples. Thank you again for your help.

**Comment by sjclare** [comment]
Hi,

I'm not sure if this fix was pushed to the main branch (1.7) but I'm getting a similar error when samples aren't diverse at an regions. I'm working with tetraploid data with regions that go up to 300 bp, but since some sample are diverse and some aren't diverse I don't want to remove the regions:

`======== Working on chromosome 'Vcev1_p0.C013'
---- Processing individual VAC_155002_P001_WG03
Number of variants skipped due to missing genotypes: 0
Number of remaining heterozygous variants: 1
Found 60 reads covering 1 variants
Kept 0 reads that cover at least two variants each
Detecting connected components with weak interconnect ..
Segmentation fault (core dumped)`

I tried installing the `issue-441` branch but since you said it merged with `issue-439` branch, I'm not sure if something happened or if I'm compiling wrong but I get this error when I try to compile it in conda:

`Installing build dependencies ... error
  error: subprocess-exited-with-error
  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> [8 lines of output]
      WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7fc75a18e620>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/setuptools/
      WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7fc75a18f430>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/setuptools/
      WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7fc75a18f6d0>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/setuptools/
      WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7fc75a18f880>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/setuptools/
      WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7fc75a18fa30>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/setuptools/
      ERROR: Could not find a version that satisfies the requirement setuptools>=60 (from versions: none)
      ERROR: No matching distribution found for setuptools>=60
      WARNING: There was an error checking the latest version of pip.
      [end of output]
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error
× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.
note: This error originates from a subprocess, and is likely not a problem with pip.
WARNING: There was an error checking the latest version of pip.`

Thanks for any help!
